qf_lib/data_providers/sp_global/sp_data_provider.py

```diff
@@ -0,0 +1,359 @@
+        to_usd: bool
+            flag indicating whether all pricing data should be converted to USD (by default False)
+
+        """
+        if frequency != Frequency.DAILY:
+            raise NotImplementedError("SPDataProvider get_history() supports currently only daily frequency.")
+
+        fields, got_single_field = convert_to_list(fields, SPField)
+        _unsupported_fields = [f for f in fields if f not in self.supported_fields]
+        if any(_unsupported_fields):
+            raise ValueError(f"Unsupported fields {_unsupported_fields}. To view the list of fields supported by "
+                             f"{self.__class__.__name__} please refer to the output of supported_fields() function.")
+
+        tickers, got_single_ticker = convert_to_list(tickers, SPTicker)
+        tid_to_tickers = {t.tradingitem_id: t for t in tickers}
+        start_datetime = self._adjust_start_date(start_datetime, frequency)
+        got_single_date = self._got_single_date(start_datetime, end_datetime, frequency)
+
+        dfs = []
+
+        # Fetch prices
+        price_fields = [f for f in fields if f in SPField.price_fields()]
+        if price_fields:
+            _prices = self._fetch_pricing_data(tid_to_tickers.keys(), price_fields, start_datetime, end_datetime,
+                                               to_usd)
+            dfs.append(_prices)
+
+        # Fetch all remaining fields
+        field_to_fetching_method = {
+            SPField.DivYield: self._fetch_dividend_yield(tid_to_tickers.keys(), start_datetime, end_datetime),
+        }
+
+        for field in [f for f in fields if f not in price_fields]:
+            try:
+                dfs.append(field_to_fetching_method[field])
+            except KeyError:
+                # TODO will be removed in the future
+                raise NotImplementedError(f"Field {field} is not supported by the SPDataProvider.") from None
+
+        grouped = pd.concat(dfs, axis=1).groupby(level=0)
+        tickers_dict = {tid_to_tickers[tid]: group.droplevel(0) for tid, group in grouped}
+        data_array = tickers_dict_to_data_array(tickers_dict, tickers, fields)
+        return normalize_data_array(data_array, tickers, fields, got_single_date, got_single_ticker, got_single_field)
+
+    def get_currency(self, tickers: Union[SPTicker, Sequence[SPTicker]]) -> Union[str, QFSeries]:
+        """
+        Function returning currency for the given selection of SPTickers.
+
+        Parameters
+        -----------
+        tickers: SPTicker, Sequence[SPTicker]
+            tickers of securities for which currencies should be retrieved
+
+        Returns
+        --------
+        str, QFSeries
+            currency for the given tickers
+        """
+
+        tickers, got_single_ticker = convert_to_list(tickers, SPTicker)
+        currencies = QFSeries()
+        sp_id_to_ticker = {t.tradingitem_id: t for t in tickers}
+
+        with self._db_connection_provider.Session.begin() as session:
```

Comment:
In some functions we have `session` as a input parameter to the function, and here we call `self._db_connection_provider.Session` - I think we should pick one approach for clarity.